env0/resource_environment.go

```diff
@@ -23,27 +23,142 @@ func resourceEnvironment() *schema.Resource {
+				Optional:    true,
+				Default:     true,
+			},
+			"deploy_on_push": {
+				Type:        schema.TypeBool,
+				Description: "should run terraform deploy on push events",
+				Optional:    true,
+				Default:     false,
+			},
+			"auto_deploy_by_custom_glob": {
+				Type: schema.TypeString,
+				// TODO: description
+				Description: "should deploy by custom glob",
+				Optional:    true,
+			},
+			"deployment_id": {
+				Type:        schema.TypeString,
+				Description: "id of the last deployment",
+				Computed:    true,
+			},
+			"configuration": {
+				Type:        schema.TypeList,
+				Description: "terraform and environment variables for the environment",
+				Optional:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
+							Type:        schema.TypeString,
+							Description: "variable name",
+							Required:    true,
+						},
+						"value": &schema.Schema{
+							Type:        schema.TypeString,
+							Description: "variable value",
+							Required:    true,
+						},
+						"type": &schema.Schema{
+							Type:        schema.TypeString,
+							Description: "variable type (allowed values are: terraform, environment)",
+							Default:     "environment",
+							Optional:    true,
+						},
+						"scope": &schema.Schema{
+							Type:        schema.TypeString,
+							Description: "variable scope ( allowed values are: GLOBAL, BLUEPRINT, PROJECT, ENVIRONMENT, DEPLOYMENT )",
+							Optional:    true,
+						},
+						"scope_id": &schema.Schema{
+							Type:        schema.TypeString,
+							Description: "the scope's id",
+							Optional:    true,
+						},
+						"description": &schema.Schema{
+							Type:        schema.TypeString,
+							Description: "description for the variable",
+							Optional:    true,
+						},
+						"is_sensitive": &schema.Schema{
+							Type:        schema.TypeBool,
+							Description: "should the variable value be hidden",
+							Optional:    true,
+						},
+						"schema_type": &schema.Schema{
+							Type:        schema.TypeString,
+							Description: "the type the variable must be of",
+							Optional:    true,
+						},
+						"schema_enum": &schema.Schema{
+							Type:        schema.TypeList,
+							Description: "a list of possible variable values",
+							Optional:    true,
+							Elem: &schema.Schema{
+								Type:        schema.TypeString,
+								Description: "name to give the configuration variable",
+							},
+						},
+					},
+				},
+			},
 		},
 	}
 }
 
+var VariableTypes = map[string]client.ConfigurationVariableType{
+	"terraform":   client.ConfigurationVariableTypeTerraform,
+	"environment": client.ConfigurationVariableTypeEnvironment,
+}
+
 func setEnvironmentSchema(d *schema.ResourceData, environment client.Environment) {
+	d.Set("id", environment.Id)
 	d.Set("name", environment.Name)
 	d.Set("project_id", environment.ProjectId)
-	d.Set("template_id", environment.TemplateId)
```

Comment:
This is important,
When a deployment happens I want to set the deployment id from the response.
If I left this here it would get overridden to latestDeploymentLog.Id on the next read ( i.e when the user runs apply\plan\refresh etc).
Now create & update set deployment id separately,
and the deployment id of an environment wont be updated if triggered from elsewhere

if you think this is a problem please bring this up.

Reply:
I think `LatestDeploymentLog` can be empty so I suggest to check if it exists first...

Reply:
9bdf19b